Improve docs markdown

ckormanyos opened the pull request:
The purpose of this PR is to add basic markdown docs

ckormanyos commented:
First draft of MD is ready for initial review(s).

Cc: @chris-durand and @rleh and @salkinium

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+The `avr-libstdcpp` port (even though not-fully-tested
+and only partially full/complete) will, nonetheless,
+be useful for those interested in making more comprehensive utilization
+of C++ --- including its standard library --- with a modern `avr-gcc` compiler.
```
```suggestion
of C++, including its standard library, with a modern `avr-gcc` compiler.
```
Long dash in Markdown isn't rendered right and looks weird

ckormanyos replied:
OK

ckormanyos replied:
Done. Will be in next commit.

salkinium commented on README.md:
````diff
@@ -1 +1,241 @@
+
+... and seamlessly use standard library headers in your code.
+
+```C
````
```suggestion
```cpp
```

ckormanyos replied:
Oh that's great. That MD detail had eluded me for toooo long. Thx

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+- `<array>` for containers having known, fixed size.
+- `<algorithm>` for standard algorithms such as sorting, minimax, sequential operations, etc.
+- `<cmath>` for projects requiring floating-point mathematical functions such as `std::sin()`, `std::exp()`, `std::frexp()` and many more.
+- `<cstdint>` which defines integral types having specified widths residing within `namespace` `std` like `std::uint8_t`.
```
```suggestion
- `<cstdint>` which defines integral types having specified widths residing within `namespace std` like `std::uint8_t`.
```

ckormanyos replied:
Done. Will be in next commit.

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+- `<cmath>` for projects requiring floating-point mathematical functions such as `std::sin()`, `std::exp()`, `std::frexp()` and many more.
+- `<cstdint>` which defines integral types having specified widths residing within `namespace` `std` like `std::uint8_t`.
+- `<limits>` offering compile-time query of numeric limits of built-in types.
+- `<numeric>` featuring a collection of useful numeric algorithms such ass `std::accumulate()`, etc.
```
```suggestion
- `<numeric>` featuring a collection of useful numeric algorithms such as `std::accumulate()`, etc.
```
butt

ckormanyos replied:
Ugggh! Where is the spell checker when you need it?

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+from a great deal of the standard library's power without compromising
+in any way on performance or sleek memory footprint.
+This is because these libaries typically require no additional storage
+and their algorithms are top-notch regarding small footprint and performance.
```
It's more that the compiler is good at optimizing them, C++ compiled with `-O0` is a nightmare

ckormanyos replied:
> the compiler is good at optimizing them

Indeed. I once wrote a passage about that in another work. The library authors actually _tune_ the way the algorith,ms are written for the characteristics of the compiler. But there is a bit more to this story. The algorithms (such as `std::sort`) are also written quite well and field tested. It would be hard with limited development time to beat those algorithms and their state of being tested.

I do not want to go into great detail _why_ it is a good idea to use the STL in the Markdown, but maybe we can find the right depth of expression together...

salkinium commented on README.md:
````diff
@@ -1 +1,241 @@
+function in an algorithmic loop with `toggle()`-method call
+expressed via lambda function.
+
+```C
````
```suggestion
```cpp
```

ckormanyos replied:
Done. Will be in next commit.

ckormanyos replied:
Done. Will be in next commit.

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+#include <array>
+#include <numeric>
+
+std::array<int, 3U> v {{ 1, 2, 3 }};
```
Why the {{}}? Is {} not enough?

ckormanyos replied:
> Why the `{{}}`? Is `{}` not enough?

That's a good queston. We might need a consensus. I'm actually a bit torn on this one myself...

I think I will actually read up on this particular programming style/artifact and get back to you on this one, as it seems quite confusing to me as well which numbers or sets of braces would be considered best...

There used to be a rather controversion pre-C++-11 debate regarding this. Somehow, I got convinced that the double-curly-braces are correct.

At the moment, I interpret `{{ 1, 2, 3, }}` to mean C++11 uniform-initialization-syntax applied to a C++11 `std::initializer_list` of `int`s. If this is right/best-practice/uniformly-so-interpreted, ... I do not know...

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+  led0,
+  led1,
+  led2
+}};
```
Same here

ckormanyos replied:
Done. Will be in next commit.

salkinium replied:
There's no need for an extra comment (we all get emails for each one), just resolve each  conversation.

salkinium commented on README.md:
````diff
@@ -1 +1,241 @@
+of items of any kind, depending on the template parameter.
+For instance,
+
+```C
````
```suggestion
```cpp
```

ckormanyos replied:
Done. Will be in next commit.

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+Consider, for instance, `std::vector` from the `<vector>` library.
+Vector creates a flexibly-sized array-like collection
+of items of any kind, depending on the template parameter.
+For instance,
```
```suggestion
For instance:
```

ckormanyos replied:
Done. Will be in next commit.

ckormanyos replied:
Done. Will be in next commit.

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+data in an operating system or mixed program/interrupt mode,
+self-written atomic primitives are needed.
+
+- ** `<random>` There is no source of entropy whatsoever on these platforms
```
```suggestion
- **`<random>`** There is no source of entropy whatsoever on these platforms
```

ckormanyos replied:
Done. Will be in next commit.

salkinium commented on README.md:
```diff
@@ -1 +1,241 @@
+to be resource-intensive and could be disruptive on tiny
+embedded platforms. An effort has been made to essentially
+remove both these library dependencies and their
+associated codes.
```
Exceptions?

salkinium reviewed (commented):
Nice!

ckormanyos commented:
> Nice!

Thanks Niklas (@salkinium). I agree with all your comments and will work them one-by-one, mabe make a few comments in the reivew areas.

ckormanyos commented:
OK folks, this one is green.
The draft is ready for reviews (if any).

Cc: @salkinium and @rleh and @chris-durand

ckormanyos commented:
OK, so I added some more details and also added an example.

This whole Markdown thing came together rather quickly, so let's give the Markdown a good second look.

If it's going in the right direction, then let's tune it. After that I can squash it all together like last PR. Anyway, guys, pls take a look, read Markdown and see what you think...

Many thanks.

salkinium commented on README.md:
```diff
@@ -1 +1,328 @@
+## Historical origins
+
+The `avr-libstdcpp` library traces its own origins to existing
+GNU/GCC C++ standard library implementaion(s) --- themselves targeting
```
```suggestion
GNU/GCC C++ standard library implementaion(s), themselves targeting
```

salkinium commented on README.md:
````diff
@@ -1 +1,328 @@
+of `avr-libstdcpp/include` on the command line...
+
+```sh
+avr-g++ -O2 -x c++ -I/my_path/avr-libstdcpp/include -mmcu=atmega328p test.cpp -o test.elf
````
```suggestion
avr-g++ -O2 -x c++ -isystem /my_path/avr-libstdcpp/include -mmcu=atmega328p test.cpp -o test.elf
```
Should this not use the system header include? It surpresses a whole lot of warnings iirc.

salkinium commented on README.md:
```diff
@@ -1 +1,328 @@
+The following non-trivial, real-world example, for instance,
+wraps instances of an overly-simplified LED class abstraction
+as object-references in an `std::array`.
+Once stored, the application exercises the leds' toggle
```
```suggestion
Once stored, the application exercises the LED's toggle
```
LE Diode's toggle, I guess that's the right use of '

salkinium commented on README.md:
```diff
@@ -1 +1,328 @@
+## Additional details
+
+`avr-libstdcpp` is intended for a modern `avr-gcc`
+such as the 11.2 port available in the [modm-io project](https://github.com/modm-io)
```
```suggestion
such as the 11.2 port available in the [modm-io project](https://github.com/modm-io/avr-gcc)
```

salkinium reviewed (approved):
I'm fine with this, just a few minor things.

chris-durand commented on README.md:
```diff
@@ -1 +1,328 @@
+## Historical origins
+
+The `avr-libstdcpp` library traces its own origins to existing
+GNU/GCC C++ standard library implementaion(s), themselves targeting
```
```suggestion
GNU/GCC C++ standard library implementation(s), themselves targeting
```

chris-durand commented on README.md:
```diff
@@ -1 +1,328 @@
+
+This vector requires storage for three integers which,
+on the `avr-gcc` platform is 6 bytes. The storage is managed
+thorugh vector's _second_, less well-known template parameter.
```
```suggestion
through vector's _second_, less well-known template parameter.
```

chris-durand commented on README.md:
````diff
@@ -1 +1,328 @@
+}
+```
+
+Using containers requires memory allocatoin with a so-called allocator.
````
```suggestion
Using containers requires memory allocation with a so-called allocator.
```

chris-durand commented on README.md:
````diff
@@ -1 +1,328 @@
+```
+
+Using containers requires memory allocatoin with a so-called allocator.
+If non is specified, as in our code snippet, the default allocator
````
```suggestion
If none is specified, as in our code snippet, the default allocator
```

chris-durand commented on README.md:
```diff
@@ -1 +1,328 @@
+- **`<chrono>`:** Only certain selected the clock functions of `<chrono>` library are implemented.
+These include `std::chrono::high_resolution_clock` and `std::chrono::steady_clock`. When using
+these clocks, it is required to implement the clock's static method
+`now()` in a project-specific fashion. The reason for this is because
+the library can not guess or independently implemtnt and actual hardware-specific
+of microcontroller-specific clock(s).
```
Something is not right in that last sentence.

ckormanyos commented:
OK guys, comprehensive commit of markdown draft now ready to go.

Many thanks for taking such detailed looks at this and finding so many typos from my side.

Cc: @chris-durand and @salkinium and @rleh

salkinium commented:
Squash?

ckormanyos commented:
> Squash?

I thought I had squashed 41 commits.

salkinium commented:
I think something went wrong, since there is a merge commit in there?

ckormanyos commented:
> think something went wrong, since there is a merge commit in there?

Maybe i wound up somewhere I don't understand. Along the way, i had merged main into the branch.

Should I try again?

ckormanyos commented:
Yes along the way, I merged main into my vranch (after we merged the `<chrono>` stuff into main.

I'm pretty sure this is good to go. I'm not sure if we work in the same way, and I#m prpbably the one not fully conformant with the flow. But I can't figure out how to consolodate this PR any more.

ckormanyos commented:
Oh. i also added a `<numeric>` example, since this was being documented in the MD. Maybe not the most brilliant way to go.

If you like, I could separate the MD changes from numeric example, roll a new, clean set of two PRs. Pls tell me how you'd like to proceed...

salkinium commented:
Thanks for the cool new Readme!

ckormanyos commented:
>  squashed and rebased your changes onto master.

Thank you.
